sdk/identity/azure-identity/src/client_certificate_credential.cpp

```diff
@@ -36,214 +47,217 @@ template <typename T> std::vector<uint8_t> ToUInt8Vector(T const& in)
-      bio = nullptr;
-
-      // Get certificate thumbprint:
-      {
-        using Azure::Core::_internal::Base64Url;
-
-        std::string thumbprintHexStr;
-        std::string thumbprintBase64Str;
+        x509.reset(PEM_read_bio_X509(bio.get(), nullptr, nullptr, nullptr));
+        if (!x509)
         {
-          std::vector<unsigned char> mdVec(EVP_MAX_MD_SIZE);
+          static_cast<void>(BIO_seek(bio.get(), 0));
+          x509.reset(PEM_read_bio_X509(bio.get(), nullptr, nullptr, nullptr));
+          if (!x509)
           {
-            unsigned int mdLen = 0;
-            const auto digestResult = X509_digest(x509, EVP_sha1(), mdVec.data(), &mdLen);
-
-            X509_free(x509);
-            x509 = nullptr;
-
-            if (!digestResult)
-            {
-              throw AuthenticationException("Failed to get certificate thumbprint.");
-            }
-
-            // Drop unused buffer space:
-            const auto mdLenSz = static_cast<decltype(mdVec)::size_type>(mdLen);
-            if (mdVec.size() > mdLenSz)
-            {
-              mdVec.resize(mdLenSz);
-            }
-
-            // Get thumbprint as hex string:
-            {
-              std::ostringstream thumbprintStream;
-              for (const auto md : mdVec)
-              {
-                thumbprintStream << std::uppercase << std::hex << std::setfill('0') << std::setw(2)
-                                 << static_cast<int>(md);
-              }
-              thumbprintHexStr = thumbprintStream.str();
-            }
+            throw AuthenticationException("Failed to read X509 section.");
           }
-
-          // Get thumbprint as Base64:
-          thumbprintBase64Str = Base64Url::Base64UrlEncode(ToUInt8Vector(mdVec));
         }
+      }
 
-        // Form a JWT token:
-        const auto tokenHeader = std::string("{\"x5t\":\"") + thumbprintBase64Str + "\",\"kid\":\""
-            + thumbprintHexStr + "\",\"alg\":\"RS256\",\"typ\":\"JWT\"}";
+      // Get certificate thumbprint:
+      unsigned int mdLen = 0;
+      const auto digestResult = X509_digest(x509.get(), EVP_sha1(), mdVec.data(), &mdLen);
 
-        const auto tokenHeaderVec
-            = std::vector<std::string::value_type>(tokenHeader.begin(), tokenHeader.end());
+      if (!digestResult)
+      {
+        throw AuthenticationException("Failed to get certificate thumbprint.");
+      }
 
-        m_tokenHeaderEncoded = Base64Url::Base64UrlEncode(ToUInt8Vector(tokenHeaderVec));
+      // Drop unused buffer space:
+      const auto mdLenSz = static_cast<decltype(mdVec)::size_type>(mdLen);
+      if (mdVec.size() > mdLenSz)
+      {
+        mdVec.resize(mdLenSz);
       }
     }
 
-    using Azure::Core::Url;
+    // Get thumbprint as hex string:
     {
-
-      m_requestUrl = Url("https://login.microsoftonline.com/");
-      m_requestUrl.AppendPath(tenantId);
-      m_requestUrl.AppendPath("oauth2/v2.0/token");
+      std::ostringstream thumbprintStream;
```

Comment:
Can we avoid using string stream here as well?

Reply:
Unlike using `std::ostringstream` simply for concatenating to strings, or for doing `std::to_string()`, here, with all these  `std::uppercase`, `std::hex`, `std::setfill('0')`, `std::setw(2)` , usage of `std::ostringstream` is justified, it handles all that formatting for us. There is no simple `to_string()` that would do all that for us using the same amount of lines, so probably we'll need to extract it to a helper function. And it is better to have the code that you don't even write, because it is less chance for bugs. This is certainly not a perf bottleneck.
We could've left stringstreams in other places either, they are not perf bottlenecks too, but if it's string concat or `to_string()`, replacement is straightforward and easy. This one - not so much.